lib/ecto/migration/auto.ex

```diff
@@ -61,14 +61,20 @@ defmodule Ecto.Migration.Auto do
       {related_field, tablename} = get_tablename(module, tablename, opts)
       tableatom = tablename |> String.to_atom
       for_opts = {related_field, opts[:for]}
-
       {fields_changes, relateds} = repo.get(SystemTable, tablename) |> Field.check(tableatom, module, for_opts)
       index_changes  = (from s in SystemTable.Index, where: ^tablename == s.tablename) |> repo.all |> Index.check(tableatom, module)
-
       if migration_module = check_gen(tableatom, module, fields_changes, index_changes, opts) do
-         Ecto.Migrator.up(repo, random, migration_module)
-         Field.update_meta(repo, module, tablename, relateds) # May be in transaction?
-         Index.update_meta(repo, module, tablename, index_changes)
+        Ecto.Migrator.up(repo, random, migration_module)
+        Field.update_meta(repo, module, tablename, relateds) # May be in transaction?
+        Index.update_meta(repo, module, tablename, index_changes)
+        case fields_changes do
```

Comment:
I think, it will better to split and move the logic for fields to `Field.update_meta` and to `Index.update_meta` for decomposing of function logic.